README.md
# Commenter Emails

A plugin for WordPress that allows for the extraction of a listing of data for all commenters (email addresses, names, URLs), and an option to export that data as a CSV file.

This plugin is available in the WordPress Plugin Directory: https://wordpress.org/plugins/commenter-emails/


## Installation

1. Install via the built-in WordPress plugin installer. Or install the plugin code inside the plugins directory for your site (typically `/wp-content/plugins/`).
2. Activate the plugin through the 'Plugins' admin menu in WordPress.
3. View the commenter email information reported in the WordPress admin via `Comments -> Commenter Emails`



## Additional Documentation

See [readme.txt](https://github.com/coffee2code/commenter-emails/blob/master/readme.txt) for additional usage information. See [CHANGELOG.md](CHANGELOG.md) for the list of changes for each release.


## Support

Commercial support and custom development are not presently available. You can raise an [issue](https://github.com/coffee2code/commenter-emails/issues) on GitHub or post in the [plugin's support forum on WordPress.org](https://wordpress.org/support/plugin/commenter-emails/).

If the plugin has been of benefit to you, how about [submitting a review](https://wordpress.org/support/plugin/commenter-emails/reviews/) for it in the WordPress Plugin Directory or considering a [donation](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=6ARCFJ9TX3522)?


## License

This plugin is free software; you can redistribute it and/or modify it under the terms of the [GNU General Public License](https://www.gnu.org/licenses/gpl-2.0.html) as published by the Free Software Foundation; either version 2 of the License, or (at your option) any later version.

CHANGELOG.md
# Changelog

## 2.6.1 _(2020-07-05)_

### Highlights:

This minor release improves markup output formatting, adds a TODO.md file, updates a few URLs to be HTTPS, expands unit testing, and notes compatibility through WP 5.4+.

### Details:

* New: Add TODO.md and move existing TODO list from top of main plugin file into it (and add to it)
* Change: Add newlines in output after block-level elements
* Change: Make an overlooked string translatable
* Change: Note compatibility through WP 5.4+
* Change: Update links to coffee2code.com to be HTTPS
* Change: Tweak code formatting
* Fix: Add a space before lists in changelog for v2.6 in readme.txt to enable Markdown list formatting to happen in Plugin Directory
* Unit tests:
    * New: Add tests for `admin_menu()`, `register_admin_menu()`
    * Change: Use HTTPS for link to WP SVN repository in bin script for configuring unit tests
    * Change: Remove unnecessary unregistering of hooks in `tearDown()`
    * Change: Add more spacing around "Tests" section separator

## 2.6 _(2020-01-01)_

### Highlights:

This minor release notes compatibility through WP 5.3+, refactors some of the code, expands unit testing, and updates the copyright date (2020).

### Details:

* Change: Extract filename handling out from `admin_menu` and into new `get_filename()`
* Change: Extract logic to determine if CSV button should be shown out from `admin_menu` and into new `should_show_csv_button()`
* Change: Extract logic to determine if email addresses should be listed on the admin page out from `admin_menu` and into new `should_show_email_addresses()`
* New: Add `get_plugin_basename()` to retrieve the plugin's basename
* Unit tests:
    * New: Add test to verify plugin hooks `plugins_loaded` action to initialize itself
    * New: Add tests for filters: `c2c_commenter_emails_filename`, `c2c_commenter_emails_show_csv_button`, `c2c_commenter_emails_show_emails`
    * New: Add tests for functions: `enqueue_admin_css()`, `plugin_action_links()`
    * New: Add approach for capturing hook values
    * Delete: Remove unnecessary `test_admin_stuff()` and its use in other tests
    * Delete: Remove duplicated test
* Change: Use `sprintf()` in places as a clearer way to construct strings rather than building them via concatenation
* Change: Note compatibility through WP 5.3+
* Change: Tweak whitespace in code and inline code docs
* Change: Update copyright date (2020)

## 2.5.1 _(2019-06-25)_
* Change: Update unit test install script and bootstrap to use latest WP unit test repo
* Change: Note compatibility through WP 5.2+

## 2.5 _(2019-04-17)_
* New: Add README.md file
* New: Add CHANGELOG.md file and move all but most recent changelog entries into it
* New: Add inline documentation for hooks
* New: Add GitHub link to readme.txt
* Change: Initialize plugin on 'plugins_loaded' action instead of on load
* Change: Merge `do_init()` into `init()`
* Change: Prevent object instantiation
    * Add private `__construct()`
    * Add private `__wakeup()`
* Change: Cast return values for a number of hooks to boolean or array
* Change: (Hardening) Encode plugin basename before use as part of a URL
* Change: Add missing translation textdomain
* Change: Unit tests: Minor whitespace tweaks to bootstrap
* Change: Note compatibility through WP 5.1+
* Change: Rename readme.txt section from 'Filters' to 'Hooks'
* Change: Modify formatting of hook name in readme to prevent being uppercased when shown in the Plugin Directory
* Change: Update copyright date (2019)
* Change: Update License URI to be HTTPS

## 2.4.1 _(2017-02-28)_
* Change: Update unit test bootstrap
    * Default `WP_TESTS_DIR` to `/tmp/wordpress-tests-lib` rather than erroring out if not defined via environment variable
    * Enable more error output for unit tests
* Change: Note compatibility through WP 4.7+
* Change: Change description
* Change: Minor readme.txt content and formatting tweaks
* Change: Update copyright date (2017)
* New: Add LICENSE file

## 2.4 _(2016-01-29)_

### Highlights:

This release introduces the ability to limit listing commenter information for specific posts, as well as many minor behind-the-scenes changes.

### Details:

* Feature: Add ability to list commenter emails for selected posts.
    * Add 'Help' panel to setting page with input field for comma-separated post IDs.
    * Add 'post_ids' arg to `get_emails()`.
    * Handle addition of 'post_ids' arg before existing 'output'.
* Change: Add support for language packs:
    * Change textdomain from 1c2c_ce' to 'commenter-emails'.
    * Don't load plugin translations from file.
    * Remove .pot file and /lang subdirectory.
    * Remove 'Domain Path' from plugin header.
* Change: Move download section below listing and make it a bit more inline.
* Change: Move initialization from `init()` into new `do_init()`, with `init()` hooking `plugins_loaded` to invoke.
* Change: Rename existing `do_init()` to `admin_menu()`.
* Change: Rename existing `admin_menu()` to `register_admin_menu()`, which is now called from `admin_menu()` instead of a hook callback.
* New: Introduce `get_post_ids()`, `help_tabs()`.
* Change: Add padding to table cells.
* Change: Make default value for 'fields' arg of `get_emails()` an empty array since function already enforces default value.
* Change: Change admin page header from 'h2' to 'h1' tag.
* Change: Minor improvements to inline docs and test docs.
* New: Create empty index.php to prevent files from being listed if web server has enabled directory listings.
* Change: Update screenshots.
* Change: Note compatibility through WP 4.4+.
* Change: Remove support for WordPress older than 4.1.
* Change: Update copyright date (2016).

## 2.3 _(2015-03-17)_
* Ensure only valid comment fields can be specified as first argument to `get_emails()`
* Use the default argument value of 'fields' argument if it is explicitly set as empty

tests/bootstrap.php
<?php
/**
 * PHPUnit bootstrap file
 *
 * @package Commenter_Emails
 */

ini_set( 'display_errors', 'on' );
error_reporting( E_ALL );

$_tests_dir = getenv( 'WP_TESTS_DIR' );
if ( ! $_tests_dir ) {
	$_tests_dir = '/tmp/wordpress-tests-lib';
}

// Give access to tests_add_filter() function.
require_once $_tests_dir . '/tests/phpunit/includes/functions.php';

/**
 * Manually load the plugin being tested.
 */
function _manually_load_plugin() {
	require dirname( dirname( __FILE__ ) ) . '/commenter-emails.php';
}
tests_add_filter( 'muplugins_loaded', '_manually_load_plugin' );

// Start up the WP testing environment.
require $_tests_dir . '/tests/phpunit/includes/bootstrap.php';


commenter-emails.php
<?php
/**
 * Plugin Name: Commenter Emails
 * Version:     2.6.1
 * Plugin URI:  https://coffee2code.com/wp-plugins/commenter-emails/
 * Author:      Scott Reilly
 * Author URI:  https://coffee2code.com/
 * Text Domain: commenter-emails
 * License:     GPLv2 or later
 * License URI: https://www.gnu.org/licenses/gpl-2.0.html
 * Description: Extract a listing of data for all commenters (email addresses, names, URLs), and an option to export that data as a CSV file.
 *
 * Compatible with WordPress 4.1+ through 5.4+.
 *
 * =>> Read the accompanying readme.txt file for instructions and documentation.
 * =>> Also, visit the plugin's homepage for additional information and updates.
 * =>> Or visit: https://wordpress.org/plugins/commenter-emails/
 *
 * @package Commenter_Emails
 * @author  Scott Reilly
 * @version 2.6.1
 */

/*
	Copyright (c) 2007-2020 by Scott Reilly (aka coffee2code)

	This program is free software; you can redistribute it and/or
	modify it under the terms of the GNU General Public License
	as published by the Free Software Foundation; either version 2
	of the License, or (at your option) any later version.

	This program is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
	GNU General Public License for more details.

	You should have received a copy of the GNU General Public License
	along with this program; if not, write to the Free Software
	Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
*/

defined( 'ABSPATH' ) or die();

if ( ! class_exists( 'c2c_CommenterEmails' ) ) :

class c2c_CommenterEmails {
	/**
	 * Stored value of plugin basename.
	 *
	 * @var string
	 * @access private
	 */
	private static $plugin_basename = '';

	/**
	 * Stored value of plugin page.
	 *
	 * @var string
	 * @access private
	 */
	private static $plugin_page     = '';

	/**
	 * Prevent instantiation.
	 *
	 * @since 2.5
	 */
	private function __construct() {}

	/**
	 * Prevent unserializing an instance.
	 *
	 * @since 2.5
	 */
	private function __wakeup() {}

	/**
	 * Returns version of the plugin.
	 *
	 * @since 2.1
	 */
	public static function version() {
		return '2.6.1';
	}

	/**
	 * Constructor.
	 */
	public static function init() {
		self::$plugin_basename = plugin_basename( __FILE__ );

		// Load textdomain.
		load_plugin_textdomain( 'commenter-emails' );

		// Register hooks.
		add_action( 'admin_menu', array( __CLASS__, 'admin_menu' ), 11 );
	}

	/**
	 * Returns the basename for the plugin's main file.
	 *
	 * @access public
	 * @since 2.6
	 *
	 * @return string
	 */
	public static function get_plugin_basename() {
		return self::$plugin_basename;
	}

	/**
	 * Determines if the button to download a CSV file of the commenter emails
	 * list should be present on the plugin's admin settings page.
	 *
	 * @access public
	 * @since 2.6
	 *
	 * @return bool True if the CSV button should be shown, else false.
	 */
	public static function should_show_csv_button() {
		/**
		 * Filters whether the button to download a CSV file of the commenter emails
		 * list should be present on the plugin's admin settings page.
		 *
		 * @since 1.2
		 *
		 * @param bool $show_csv_button Show the CSV button? Default true.
		 */
		return (bool) apply_filters( 'c2c_commenter_emails_show_csv_button', true );
	}

	/**
	 * Determines if the listing of email addresses should appear on the plugin's
	 * admin settings page.
	 *
	 * @access public
	 * @since 2.6
	 *
	 * @return bool True if the listing of email addresses should be shown, else false.
	 */
	public static function should_show_email_addresses() {
		/**
		 * Filters whether the listing of email addresses should appear on the
		 * plugin's admin settings page.
		 *
		 * @since 1.2
		 *
		 * @param bool $show_emails Show the listing of email addresses? Default true.
		 */
		return (bool) apply_filters( 'c2c_commenter_emails_show_emails', true );
	}

	/**
	 * Returns the name to be used for the CSV file.
	 *
	 * Uses `c2c_commenter_emails_filename` filter to customize filename.
	 *
	 * Note: Does not attempt to ensure filtered filename is sane.
	 *
	 * @access public
	 * @since 2.6
	 *
	 * @return string
	 */
	public static function get_filename() {
		/**
		 * Filters the name used for the .csv file when being downloaded.
		 *
		 * @since 1.2
		 *
		 * @param string $filename The filename. Default 'commenter-emails-YYYY-MM-DD-H:S.csv'.
		 */
		return apply_filters(
			'c2c_commenter_emails_filename',
			sprintf( 'commenter-emails-%s.csv',  mysql2date( 'Y-m-d-Hi', current_time( 'mysql' ) ) )
		);
	}

	/**
	 * Initialize hooks and data.
	 */
	public static function admin_menu() {
		self::register_admin_menu();

		if ( self::$plugin_page ) {
			// Handles CSV download.
			add_action( 'load-' . self::$plugin_page, array( __CLASS__, 'handle_csv_download' ) );
			// Register and enqueue styles for admin page.
			add_action( 'load-' . self::$plugin_page, array( __CLASS__, 'enqueue_admin_css' ) );
			// Register help tabs.
			add_action( 'load-' . self::$plugin_page, array( __CLASS__, 'help_tabs' ) );
		}
	}

	/**
	 * Query database to obtain the list of commenter email addresses.
	 *
	 * Only checks comments that are approved, have a author email, and are
	 * of the comment_type 'comment' (or '').
	 *
	 * Only one entry is returned per email address. If a given email address
	 * has multiple instances in the database, each with different names, then
	 * the most recent comment will be used to obtain any additional field data
	 * such as comment_author, etc.
	 *
	 * @param array  $fields   Optional. The fields to obtain from each comment.
	 * @param array  $post_ids Optional. IDs of posts whose comments should be considered.
	 * @param string $output   Optional. Any of ARRAY_A | ARRAY_N | OBJECT | OBJECT_K
	 *                         constants. See WP docs for wpdb::get_results() for
	 *                         more info.
	 * @return mixed List of email addresses.
	 */
	public static function get_emails( $fields = array(), $post_ids = array(), $output = ARRAY_N ) {
		global $wpdb;

		// If $field is explicitly empty, use the default.
		if ( ! $fields ) {
			$fields = array( 'comment_author_email', 'comment_author', 'comment_author_url' );
		}

		// Ensure only valid comment fields are specified.
		$fields = array_intersect(
			(array) $fields,
			array( 'comment_ID', 'comment_post_ID', 'comment_author', 'comment_author_email', 'comment_author_url', 'comment_author_IP',
					'comment_date', 'comment_date_gmt', 'comment_content', 'comment_karma', 'comment_approved', 'comment_agent',
					'comment_type', 'comment_parent', 'user_id' )
		);

		// In v2.4, the $post_ids arg was introduced and put before the existing
		// $output arg. This block of code handles backwards compatibility for whoever
		// may be using that arg.
		if ( ! is_array( $post_ids ) && in_array( $post_ids, array( ARRAY_A, ARRAY_N, OBJECT, OBJECT_K ) ) ) {
			$output = $post_ids;
			$post_ids = array();
		}

		// Ensure comment_author_email is always included.
		if ( ! in_array( 'comment_author_email', $fields ) ) {
			array_unshift( $fields,  'comment_author_email' );
		}

		$fields = implode( ', ', $fields );

		if ( $post_ids ) {
			$post_ids = implode( ', ', array_map( 'intval', (array) $post_ids ) );
			$posts = " AND comment_post_ID IN ( $post_ids ) ";
		} else {
			$posts = '';
		}

		$sql = "SELECT $fields
				FROM {$wpdb->comments} t1
				INNER JOIN ( SELECT MAX(comment_ID) AS id FROM {$wpdb->comments} GROUP BY comment_author_email ) t2 ON t1.comment_ID = t2.id
				WHERE
					comment_approved = '1' AND
					comment_author_email != '' AND
					(comment_type = '' OR comment_type = 'comment')
					{$posts}
				GROUP BY comment_author_email
				ORDER BY comment_author_email ASC";
		$emails = $wpdb->get_results( $sql, $output );

		return $emails;
	}

	/**
	 * Handles download of commenter emails directly as CSV file.
	 */
	public static function handle_csv_download() {
		$csv_filename = self::get_filename();

		if ( isset( $_GET['download_csv'] ) && '1' == $_GET['download_csv'] ) {
			header( 'Content-type: text/csv' );
			header( 'Cache-Control: no-store, no-cache' );
			header( 'Content-Disposition: attachment; filename="' . $csv_filename . '"' );

			$outstream = fopen( "php://output", 'w' );

			$default_fields = array( 'comment_author', 'comment_author_email' );
			if ( isset( $_GET['include_url'] ) && '1' == $_GET['include_url'] ) {
				$default_fields[] = 'comment_author_url';
			}

			/**
			 * Filters the user fields included in the download CSV file.
			 *
			 * @since 2.0
			 *
			 * @param array $fields The user fields. Whether explicitly included or not,
			 *                      'comment_author_email' will always be output in the
			 *                      CSV. Default ['comment_author', 'comment_author_email'].
			 */
			$fields = (array) apply_filters( 'c2c_commenter_emails_fields', $default_fields );

			/**
			 * Filters the user fields included in the download CSV file. By default the
			 * CSV file includes comment_author and comment_author_email.
			 *
			 * @since 2.0
			 *
			 * @param string $separator String to be used as the data separator in the
			 *                          CSV file. Default is ','.
			 */
			$field_sep = apply_filters( 'c2c_commenter_emails_field_separator', ',' );

			foreach ( (array) self::get_emails( $fields, self::get_post_ids() ) as $item ) {
				fputcsv( $outstream, $item, $field_sep, '"' );
			}

			fclose( $outstream );

			exit();
		}
	}

	/**
	 * Creates the admin menu.
	 */
	public static function register_admin_menu() {
		// Add plugin action links.
		add_filter( 'plugin_action_links_' . self::$plugin_basename, array( __CLASS__, 'plugin_action_links' ) );

		/**
		 * Filters the capability required to access the commenter emails admin page.
		 *
		 * Ensure capability is created and assigned to the desired user(s)
		 *
		 * @since 1.1.1
		 *
		 * @param string $capability The capability name. Default 'manage_options'.
		 */
		$capability = apply_filters( 'manage_commenter_emails_options', 'manage_options' );

		// Add menu item under Comments.
		self::$plugin_page = add_comments_page(
			__( 'Commenter Emails', 'commenter-emails' ),
			__( 'Commenter Emails', 'commenter-emails' ),
			$capability,
			self::$plugin_basename,
			array( __CLASS__, 'admin_page' )
		);
	}

	/**
	 * Enqueues stylesheets.
	 *
	 * @since 2.1
	 */
	public static function enqueue_admin_css() {
		wp_register_style( __CLASS__, plugins_url( 'commenter-emails.css', __FILE__ ) );
		wp_enqueue_style( __CLASS__ );
	}

	/**
	 * Adds a 'Settings' link to the plugin action links.
	 *
	 * @param array  $action_links The current action links.
	 * @return array The action links.
	 */
	public static function plugin_action_links( $action_links ) {
		$settings_link = sprintf(
			'<a href="%s">%s</a>',
			esc_url( 'edit-comments.php?page=' . urlencode( self::$plugin_basename ) ),
			__( 'Listing', 'commenter-emails' )
		);
		array_unshift( $action_links, $settings_link );
		return $action_links;
	}

	/**
	 * Gets the post IDs specified by the user.
	 *
	 * @since 2.4
	 *
	 * @return array
	 */
	protected static function get_post_ids() {
		$key = 'c2c-commenter-emails-post-ids';
		if ( ! isset( $_GET[ $key ] ) || ! $_GET[ $key ] ) {
			return array();
		}

		return array_unique( array_map( 'intval', array_map( 'trim', explode( ',', $_GET[ $key ] ) ) ) );
	}

	/**
	 * Outputs the contents of the plugin's admin page.
	 */
	public static function admin_page() {
		$post_ids = self::get_post_ids();
		$emails = self::get_emails( array(), $post_ids );
		$emails_count = count( $emails );
		$logo = plugins_url( 'c2c_minilogo.png', __FILE__ );

		echo '<div class="wrap">';
		echo '<h1>' . __( 'Commenter Emails', 'commenter-emails' ) . '</h1>';
		echo '<p>';
		if ( $post_ids ) {
			printf(
				/* translators: %s: count of number of commenter email addresses */
				_n(
					'There is %s unique and approved commenter email address for the specified posts.',
					'There are %s unique and approved commenter email addresses for the specified posts.',
					$emails_count,
					'commenter-emails'
				),
				$emails_count
			);
		} else {
			printf(
				/* translators: %s: count of number of commenter email addresses */
				_n(
					'There is %s unique and approved commenter email address for this site.',
					'There are %s unique and approved commenter email addresses for this site.',
					$emails_count,
					'commenter-emails'
				),
				$emails_count
			);
		}
		echo '</p>';
		echo '</div>';

		if ( self::should_show_email_addresses() ) {
			echo '<div class="wrap">';
			echo '<h2>' . __( 'All Commenter Emails', 'commenter-emails' ) . "</h2>\n";

			if ( $emails ) {
				echo '<table class="commenter-emails-table">';
				echo '<tr><th>' . __( 'Email', 'commenter-emails' ) . '</th><th>' . __( 'Name', 'commenter-emails' ) . '</th>';
				echo '<th>' . __( 'URL', 'commenter-emails' ) . '</th></tr>';

				foreach ( $emails as $item ) {
					echo '<tr><td>' . esc_html( $item[0] ) . '</td><td>' . esc_html( $item[1] ) . '</td>';
					echo '<td>' . make_clickable( esc_html( $item[2] ) ) . '</td></tr>';
				}

				echo "</table>\n";
			}

			echo '<p>';
			if ( $post_ids ) {
				if ( count( $post_ids ) > 1 ) {
					if ( $emails_count ) {
						printf( _n( '%s commenter email listed for the specified posts.', '%s commenter emails listed for the specified posts.', $emails_count, 'commenter-emails' ), $emails_count );
					} else {
						_e( 'There were no commenters found to list for the specified posts.', 'commenter-emails' );
					}
				} else {
					if ( $emails_count ) {
						printf( _n( '%s commenter email listed for the specified post.', '%s commenter emails listed for the specified post.', $emails_count, 'commenter-emails' ), $emails_count );
					} else {
						_e( 'There were no commenters found to list for the specified post.', 'commenter-emails' );
					}
				}
			} elseif ( ! $emails_count ) {
				printf( __( 'There were no commenters found to list.', 'commenter-emails' ) );
			} else {
				printf( __( '%s commenter emails listed.', 'commenter-emails' ), $emails_count );
			}
			echo "</p>\n";
			echo "</div>\m";
		}

		if ( self::should_show_csv_button() ) {
			echo '<div class="wrap">';
			echo '<p><form action="" method="get">';
			echo '<label for="submit">';
			_e( 'Download this list of email addresses as a CSV file :', 'commenter-emails' );
			echo ' <input type="submit" name="submit" value="' . esc_attr__( 'Download', 'commenter-emails' ) . '" />';
			echo '</label>';
			echo '<label for="include_url">(';
			echo '<input type="checkbox" name="include_url" value="1" /> ';
			_e( 'Include commenter website?', 'commenter-emails' );
			echo ')<input type="hidden" name="page" value="' . esc_attr( $_GET['page'] ) . '" />';
			echo '<input type="hidden" name="download_csv" value="1" />';
			echo "</form></p></div>\n";
		}

		echo '<div class="wrap">';
		echo '<p class="description">';
		_e( 'NOTE: If you would like to list only commenters for selected posts, use the "Help" tab above to specify those posts.', 'commenter-emails' );
		echo '</p>';
		echo "</div>\n";

		echo '<div id="c2c-ce" class="wrap"><div>';
		printf( __( 'This plugin brought to you by <a href="%s">Scott Reilly, aka coffee2code</a>.', 'commenter-emails' ), 'https://coffee2code.com' );
		echo '<span><a href="https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=6ARCFJ9TX3522" title="' . esc_attr__( 'Please consider a donation', 'commenter-emails' ) . '">';
		_e( 'Did you find this plugin useful?', 'commenter-emails' );
		echo "</a></span></div></div>\n";
	}

	/**
	 * Registers help tabs.
	 *
	 * @since 2.4
	 */
	public static function help_tabs() {
		if ( ! class_exists( 'WP_Screen' ) ) {
			return;
		}

		$screen = get_current_screen();

		if ( $screen->id !== self::$plugin_page ) {
			return;
		}

		$content = '<form action="" method="get">';
		$content .= '<p><label for="c2c-commenter-emails-post-ids">';
		$content .=  __( 'Filter the listing to only include commenters who commented on specified posts.', 'commenter-emails' );
		$content .= '</label></p>';
		$content .= '<input type="text" size="50" name="c2c-commenter-emails-post-ids" value="' . implode( ', ', self::get_post_ids() ) . '"/>' . "\n";
		$content .= '<p class="description">';
		$content .= __( 'Comma-separated list of post IDs to limit commenter emails to those posts.', 'commenter-emails' );
		$content .= "</p>\n";
		$content .= '<input type="hidden" name="page" value="' . esc_attr( $_GET['page'] ) . '"/>';
		$content .= '<input type="submit" name="c2c-commenter-emails-apply" class="button" value="' . esc_attr__( 'Apply', 'commenter-emails' ) . '" />';
		$content .= "</form>\n";

		$screen->add_help_tab( array(
			'id'      => 'c2c-commenter-emails',
			'title'   => __( 'Limit to posts', 'commenter-emails' ),
			'content' => $content,
		) );
	}
} // end c2c_CommenterEmails

add_action( 'plugins_loaded', array( 'c2c_CommenterEmails', 'init' ) );

endif; // end if ! class_exists()
